File: educational_dp/ruby/g.rb
# G
# トポロジカルソート

N, M = gets.split.map(&:to_i)
G = Array.new(N) { [] }
# 有効辺の向けられた数
arrows = [0] * N

M.times do
  x, y = gets.split.map(&:to_i).map(&:pred)
  G[x] << y
  arrows[y] += 1
end

# 有効辺の向けられた数が0なのは始点となるもの
next_candicates = N.times.filter { arrows[_1].zero? }

depth = [0] * N
until next_candicates.empty?
  pos = next_candicates.shift
  G[pos].each do |i|
    arrows[i] -= 1
    next_candicates << i if arrows[i].zero?
    depth[i] = [depth[i], depth[pos] + 1].max
  end
end

puts depth.max

File: educational_dp/kotlin/g.kt
// G

fun main() {
  val (N, M) = readLine()!!.split(" ").map(String::toInt)
  val G = MutableList(N) { mutableListOf<Int>() }

  val arrows = IntArray(N) { 0 }
  repeat(M) {
    val (x, y) = readLine()!!.split(" ").map(String::toInt).map(Int::dec)
    G[x].add(y)
    arrows[y]++
  }

  var candidates = (0 .. N - 1).filter { arrows[it] == 0 }.toMutableList()
  val depth = IntArray(N) { 0 }

  while (!candidates.isEmpty()) {
    val pos = candidates.removeAt(0)
    G[pos].forEach {
      if (--arrows[it] == 0) candidates.add(it)
      depth[it] = Math.max(depth[it], depth[pos] + 1)
    }
  }

  println(depth.max())
}
